File: frontend/src/lib/queue.ts
import { post, request } from "@/lib/api";
import type {
  AppointmentDetail,
  AppointmentType,
  DoctorRef,
  PatientRef,
  Status,
} from "@/lib/appointments";

export type QueueStatus =
  "waiting" | "called" | "in_consultation" | "completed" | "skipped" | "no_show";

export type Priority = "normal" | "urgent";

export type QueueEntry = {
  id: string;
  token: number;
  status: QueueStatus;
  priority: Priority;
  patient: PatientRef;
  doctor: DoctorRef;
  appointment: {
    id: string;
    start_time: string;
    end_time: string;
    appointment_type: AppointmentType;
    reason: string | null;
    status: Status;
  } | null;
  reason: string | null;
  checked_in_at: string;
  called_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  /** Worked out on the server, whose clock the timestamps came from. */
  waited_minutes: number;
  position: number | null;
  expected_wait_minutes: number | null;
};

export type Arrival = {
  id: string;
  patient: PatientRef;
  start_time: string;
  end_time: string;
  appointment_type: AppointmentType;
  status: Status;
  reason: string | null;
  is_late: boolean;
};

export type Lane = {
  doctor: DoctorRef;
  closed: string | null;
  now_seeing: QueueEntry | null;
  called: QueueEntry | null;
  waiting: QueueEntry[];
  skipped: QueueEntry[];
  done: QueueEntry[];
  expected: Arrival[];
  seen_count: number;
  average_minutes: number;
};

export type QueueDay = {
  date: string;
  day_name: string;
  lanes: Lane[];
  only_doctor_id: string | null;
  unlinked: boolean;
};

export type Step = "call" | "start" | "complete" | "skip" | "recall" | "no-show";

/** How often a screen showing the queue asks again, while it is in view. */
export const POLL_MS = 5_000;

export const getQueue = (doctor?: string) =>
  request<QueueDay>(`/queue${doctor ? `?doctor_id=${encodeURIComponent(doctor)}` : ""}`);

export const checkIn = (appointmentId: string, priority: Priority = "normal") =>
  post<QueueEntry>(`/appointments/${appointmentId}/check-in`, { priority });

export const addWalkIn = (walkIn: {
  patient_id: string;
  doctor_id: string;
  reason?: string | null;
  priority: Priority;
}) => post<QueueEntry>("/queue/walk-in", walkIn);

export const takeStep = (entryId: string, step: Step) =>
  post<QueueEntry>(`/queue/${entryId}/${step}`, {});

export const setPriority = (entryId: string, priority: Priority) =>
  request<QueueEntry>(`/queue/${entryId}`, {
    method: "PATCH",
    body: JSON.stringify({ priority }),
  });

/** Takes back a check-in. A booked patient's appointment comes back as it was. */
export const undoCheckIn = (entryId: string) =>
  request<AppointmentDetail | null>(`/queue/${entryId}`, { method: "DELETE" });

/** "12 min", "1 hr 5 min": a wait, the way somebody at the desk says it. */
export function spokenMinutes(minutes: number): string {
  if (minutes < 1) return "under a minute";
  if (minutes < 60) return `${minutes} min`;
  const hours = Math.floor(minutes / 60);
  const left = minutes % 60;
  return left ? `${hours} hr ${left} min` : `${hours} hr`;
}

/** A guess, and said like one. Nobody should be promised a minute. */
export function roughly(minutes: number | null): string | null {
  if (minutes === null) return null;
  if (minutes <= 0) return "Next in line";
  const rounded =
    minutes < 15 ? Math.max(5, Math.round(minutes / 5) * 5) : Math.round(minutes / 10) * 10;
  return `In about ${spokenMinutes(rounded)}`;
}

/** The name somebody is called by, which is not always the one on file. */
export function calledBy(patient: PatientRef): string {
  return patient.preferred_name ?? patient.full_name.split(" ")[0] ?? patient.full_name;
}

File: backend/app/schemas/queue.py
"""Request and response shapes for the day's queue."""

from __future__ import annotations

import datetime as dt
import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.appointment import AppointmentType, DoctorRef, PatientRef, Status

QueueStatus = Literal["waiting", "called", "in_consultation", "completed", "skipped", "no_show"]
Priority = Literal["normal", "urgent"]


class _Trimmed(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)


class WalkIn(_Trimmed):
    patient_id: uuid.UUID
    doctor_id: uuid.UUID
    reason: str | None = Field(default=None, max_length=200)
    priority: Priority = "normal"


class CheckIn(_Trimmed):
    priority: Priority = "normal"


class PriorityWrite(BaseModel):
    priority: Priority


class BookedFor(BaseModel):
    """The appointment a place came from, for the line that says so."""

    id: uuid.UUID
    start_time: dt.time
    end_time: dt.time
    appointment_type: AppointmentType
    reason: str | None
    status: Status


class VitalsBrief(BaseModel):
    """What the queue shows of the readings taken for a place."""

    id: uuid.UUID
    systolic_mmhg: int | None
    diastolic_mmhg: int | None
    pulse_bpm: int | None
    temperature_c: float | None
    spo2_percent: int | None
    weight_kg: float | None
    glucose_mg_dl: int | None
    flags: dict[str, Literal["high", "low"]]


class QueueEntryOut(BaseModel):
    id: uuid.UUID
    token: int
    status: QueueStatus
    priority: Priority
    patient: PatientRef
    doctor: DoctorRef
    appointment: BookedFor | None
    reason: str | None
    checked_in_at: dt.datetime
    called_at: dt.datetime | None
    started_at: dt.datetime | None
    completed_at: dt.datetime | None
    # Minutes since they arrived while they wait, and how long they waited
    # once they have gone in. Worked out on the server, whose clock the
    # timestamps came from.
    waited_minutes: int
    # Where they stand in the line, counting from one, while they wait.
    position: int | None
    # A guess from how long this doctor's consultations have been taking
    # today. Never stored, and never shown as more than roughly.
    expected_wait_minutes: int | None
    # The notes written for this place, once the doctor has opened them.
    consultation_id: uuid.UUID | None = None
    # The prescription issued at this visit, for the desk to print.
    prescription_id: uuid.UUID | None = None
    vitals: VitalsBrief | None = None


class Arrival(BaseModel):
    """An appointment today that has not been checked in yet."""

    id: uuid.UUID
    patient: PatientRef
    start_time: dt.time
    end_time: dt.time
    appointment_type: AppointmentType
    status: Status
    reason: str | None
    is_late: bool


class Lane(BaseModel):
    doctor: DoctorRef
    # Why nobody can join this doctor's line today, if nobody can.
    closed: str | None
    now_seeing: QueueEntryOut | None
    called: QueueEntryOut | None
    waiting: list[QueueEntryOut]
    skipped: list[QueueEntryOut]
    done: list[QueueEntryOut]
    expected: list[Arrival]
    seen_count: int
    average_minutes: int


class QueueDay(BaseModel):
    date: dt.date
    day_name: str
    lanes: list[Lane]
    only_doctor_id: uuid.UUID | None
    unlinked: bool
